mobile/ios/ios-core.ts

```diff
@@ -797,51 +809,74 @@ class PosixSocket implements Mobile.IiOSDeviceSocket {
 
 	public receiveMessage(): IFuture<Mobile.IiOSSocketResponseData> {
 		var result = new Future<Mobile.IiOSSocketResponseData>();
-		var capturedData: NodeBuffer = new Buffer(0);
 
 		this.socket
 			.on("data", (data: NodeBuffer) => {
-				capturedData = Buffer.concat([capturedData, data]);
-
-				if(this.format === CoreTypes.kCFPropertyListBinaryFormat_v1_0) {
-					var isExceptionThrown = false;
-
+				this.buffer = Buffer.concat([this.buffer, data]);
+				if (this.format === CoreTypes.kCFPropertyListBinaryFormat_v1_0) {
 					try {
-						var message = bplistParser.parseBuffer(data);
-					} catch(e) {
-						isExceptionThrown = true;
-					}
-
-					if(!isExceptionThrown) {
-						this.$logger.trace("MESSAGE RECEIVING");
-						this.$logger.trace(message);
-
-						if(message && typeof(message) === "object" && message[0]) {
-							message = message[0];
-							var output = "";
-							if(message.Status) {
-								output += util.format("Status: %s", message.Status);
-							}
-							if(message.PercentComplete) {
-								output += util.format(" PercentComplete: %s", message.PercentComplete);
-							}
-							this.$logger.out(output);
-
-							if(message.Status && message.Status === "Complete") {
-								if(!result.isResolved()) {
-									result.return(message);
-								}
+						while (this.buffer.length >= this.length) {
+							switch (this.state) {
+								case ReadState.Length:
+									this.length = this.buffer.readInt32BE(0);
+									this.buffer = this.buffer.slice(4);
+									this.state = ReadState.Plist;
+									break;
+								case ReadState.Plist:
+									try {
+										var plistBuffer = this.buffer.slice(0, this.length);
+										var message = bplistParser.parseBuffer(plistBuffer);
+										this.$logger.trace("MESSAGE RECEIVING");
+										this.$logger.trace(message);
+										try {
+											if (message && typeof (message) === "object" && message[0]) {
```

Comment:
I've pretty much preserved this code. I do not want to know what the message is.